Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.cpp

```diff
@@ -3384,24 +3386,33 @@ static ImageOrientation getVideoOrientation(const GstTagList* tagList)
     if (!g_strcmp0(tag.get(), "rotate-270"))
         return ImageOrientation::OriginLeftBottom;
 
-    // Default rotation.
-    return ImageOrientation::None;
+    return std::nullopt;
 }
 
 void MediaPlayerPrivateGStreamer::updateVideoOrientation(const GstTagList* tagList)
 {
+    // We're in videoSinkPad streaming thread.
+    ASSERT(!isMainThread());
     GST_DEBUG_OBJECT(pipeline(), "Updating orientation from %" GST_PTR_FORMAT, tagList);
-    auto sizeActuallyChanged = setVideoSourceOrientation(getVideoOrientation(tagList));
+    auto orientation = getVideoOrientation(tagList);
+    if (!orientation)
+        return;
 
-    if (!sizeActuallyChanged)
+    if (!setVideoSourceOrientation(*orientation))
         return;
 
+    auto previousVideoSize = m_videoSize;
+
     // If the video is tagged as rotated 90 or 270 degrees, swap width and height.
-    if (m_videoSourceOrientation.usesWidthAsHeight())
-        m_videoSize = m_videoSize.transposedSize();
+    auto newVideoSize = m_videoSourceOrientation.usesWidthAsHeight() ? m_videoSize.transposedSize() : previousVideoSize;
+
+    // Prevent false positive sizeChanged events.
+    if (newVideoSize == previousVideoSize)
+        return;
 
     GST_DEBUG_OBJECT(pipeline(), "Enqueuing and waiting for main-thread task to call sizeChanged()...");
-    bool sizeChangedProcessed = m_sinkTaskQueue.enqueueTaskAndWait<AbortableTaskQueue::Void>([this] {
+    bool sizeChangedProcessed = m_sinkTaskQueue.enqueueTaskAndWait<AbortableTaskQueue::Void>([this, newVideoSize] {
+        m_videoSize = newVideoSize;
```

Comment:
Since we don't keep the unrotated video size, this would still fail in Case 2:

1. "rotate-90" is received, m_videoSize is transposed, getting vertical video (correct).
2. "rotate-270" is received, m_videoSize is wrongly transposed again, resulting in playing vertical video as horizontal video.

Reply:
The rotation is relative to the current video frame dimensions, not any previous video frame dimensions. Not sure to see what's the problem here, TBH.

Reply:
> The rotation is relative to the current video frame dimensions

Wouldn't it be relative to the size in the caps?

Reply:
That's another way of saying it, yes.

Reply:
Do you have the guarantee that if a second image rotation tag event occurs, m_videoSize will still have the unrotated size from the caps, and not the rotated size from the previous image rotation event?

Reply:
The tag event is serialized with data flow. Isn't that enough?

Reply:
Imagine you get the following events, serialized:

1. caps, video/[...] width=640, height=480
2. tag, image-orientation=rotate-90
3. tag, image-orientation=rotate-270

Reply:
The computed size would be 480x640 after the first tag, and back to 640x480 after the second event.

Reply:
Therefore, after the second tag event it would be using a horizontal size despite the content being vertical.

Reply:
I'm updating the patch. Thanks!